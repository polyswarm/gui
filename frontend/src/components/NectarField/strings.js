const strings = {
  nectar: 'Nectar (NCT)'
};

export default strings;