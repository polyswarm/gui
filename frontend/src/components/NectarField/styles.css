.NectarField {
    display: flex;
    flex-direction: column;
}

.NectarField-Label {
    color: #a5a5a5;
    line-height: 36px;
    position: relative;
    user-select: none;
    top: 42px;
}

.NectarField-Input {
    background-color: rgba(0, 0, 0, 0);
    border: none;
    border-bottom: white 2px solid;
    color: white;
    box-sizing: border-box;
    font-size: 24px;
    height: 48px;
    line-height: 48px;
}

.NectarField-Input:focus {
    border-bottom: rgb(46, 255, 109) 2px solid;
    outline-color: rgba(0,0,0,0);
}

input[type=number]::-webkit-inner-spin-button, 
input[type=number]::-webkit-outer-spin-button { 
  -webkit-appearance: none; 
  margin: 0; 
}

.error {
    border-bottom: red 2px solid;
}

.error:focus {
    border-bottom: red 2px solid;
    outline-color: rgba(0,0,0,0);
}

.label-enter {
    color: #a5a5a5;
    font-size: 36px;
    font-weight: normal;
    transform: translateY(0px);
}
.label-enter-active {
    color: rgb(46, 255, 109);
    font-size: 24px;
    font-weight: 300;
    transform: translateY(-36px);
    transition: all 300ms ease-out;
}

.label-enter-done {
    color: rgb(46, 255, 109);
    font-size: 24px;
    font-weight: 300;
    transform: translateY(-36px);
}

.label-exit {
    color: rgb(46, 255, 109);
    font-size: 24px;
    font-weight: 300;
    transform: translateY(-36px);
}
.label-exit-active {
    color: #a5a5a5;
    font-size: 36px;
    font-weight: normal;
    transform: translateY(0px);
    transition: all 300ms ease-out;
}

.label-exit-done {
    color: #a5a5a5;
    font-size: 36px;
    font-weight: normal;
    transform: translateY(0px);
}