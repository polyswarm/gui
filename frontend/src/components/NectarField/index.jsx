// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {CSSTransition} from 'react-transition-group';
// Project imports
// Component imports
import strings from './strings';
import './styles.css';

class NectarField extends Component {
  constructor(props) {
    super(props);
    this.state = {
      focused: false,
      error: false,
      first: true,
    }

    this.onBlur = this.onBlur.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.getNectar = this.getNectar.bind(this);
  }

  componentDidMount() {
    this.setState({first: false});
  }

  render() {
    const {state: {error, focused} } = this;
    const nectar = this.getNectar();
    const inputClass = classNames('NectarField-Input', {error: error});
    return (
        <div className='NectarField'>
          <CSSTransition
            in={(nectar && nectar.length > 0) || focused}
            timeout={300}
            classNames='label'> 
            {() => (
              <label className='NectarField-Label'
                htmlFor='nectar'>
                {strings.nectar}
              </label>
            )}
          </CSSTransition>
          <input
            className={inputClass}
            id='nectar'
            type='number'
            value={nectar}
            onBlur={this.onBlur}
            onFocus={this.onFocus}
            onChange={this.onChange}/>
        </div>
    );
  }

  onBlur() {
    this.setState({focused: false});
  }

  onChange(e) {
    const {props: {onChange}} = this;
    const nectar = e.target.value;
    const valid = this.validateNectar(nectar)
    if (onChange) {
      onChange(nectar, valid);
    }
    this.setState({error: !valid});
  }

  onFocus() {
    this.setState({focused: true});
  }

  /**
   * This triggers a redraw of the element so we get CSSTransition to animate
   * when the nectar is set as a prop for the first draw.
   */
  getNectar() {
    const { state: {first}, props: {nectar}} = this;
    if (first) {
      return '';
    } else {
      return nectar;
    }
  }

  validateNectar(nectar) {
    return nectar.length == 0 || nectar.length > 14;
  }
}
NectarField.proptypes = {
  onChange: PropTypes.func,
  nectar: PropTypes.string
}
export default NectarField;