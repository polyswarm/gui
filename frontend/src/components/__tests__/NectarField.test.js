import React from 'react';
import {render, shallow} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import NectarField from '../NectarField';

it('renders without crashing', () => {
  const wrapper = render(<NectarField />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('calls onchange when instance onChange is called', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<NectarField onChange={onChange}/>);
  const instance = wrapper.instance();

  instance.onChange({target:{value: 'change'}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith('change', false);
});

it('calls onchange when nectar changed', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<NectarField onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: 'change'}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith('change', false);
});

it('calls onChange with valid:true when value is over 0 ', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<NectarField onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: .0000000000000001}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith(.0000000000000001, true);
});

it('calls onchange with valid:false when empty', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<NectarField onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: ''}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith('', false);
});

it('calls onchange with valid:false value is 0', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<NectarField onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: 0}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith(0, false);
});

it('calls onchange with valid:false value is less than 0', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<NectarField onChange={onChange}/>);
  
  wrapper.find('input').simulate('change', {target: {value: -1}});

  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith(-1, false);
});

it('sets error to true when set nectar is 0', () => {
  const setState = jest.spyOn(NectarField.prototype, 'setState');
  const wrapper = shallow(<NectarField />);
  setState.mockClear();
  
  wrapper.find('input').simulate('change', {target: {value: 0}});

  expect(setState).toHaveBeenCalledWith({error: true});
});

it('sets error to true when set nectar is less than 0', () => {
  const setState = jest.spyOn(NectarField.prototype, 'setState');
  const wrapper = shallow(<NectarField />);
  setState.mockClear();
  
  wrapper.find('input').simulate('change', {target: {value: -1}});

  expect(setState).toHaveBeenCalledWith({error: true});
});

it('sets error to false when set nectar is greater than 0', () => {
  const setState = jest.spyOn(NectarField.prototype, 'setState');
  const wrapper = shallow(<NectarField />);
  setState.mockClear();
  
  wrapper.find('input').simulate('change', {target: {value: '1'}});

  expect(setState).toHaveBeenCalledWith({error: false});
});