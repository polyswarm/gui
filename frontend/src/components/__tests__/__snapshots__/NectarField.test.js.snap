// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="NectarField"
>
  <label
    class="NectarField-Label"
    for="nectar"
  >
    Enter nectar
  </label>
  <input
    class="NectarField-Input"
    id="nectar"
    type="number"
    value=""
  >
    
  </input>
</div>
`;
